src/strbuf.c

```diff
@@ -0,0 +1,483 @@
+    }
+
+    *success = 1;
+    return ret;
+}
+
+int Strbuf_Compare(const Strbuf *str1, const Strbuf *str2)
+{
+    Strbuf_Check(str1);
+    Strbuf_Check(str2);
+
+    for (int i = 0; str1->data[i] == str2->data[i]; i++) {
+        if (str1->data[i] == EOS) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+u32 Strbuf_Length(const Strbuf *strbuf)
+{
+    Strbuf_Check(strbuf);
+    return strbuf->size;
+}
+
+u32 Strbuf_NumLines(const Strbuf *strbuf)
+{
+    Strbuf_Check(strbuf);
+
+    int i, count;
+    for (i = 0, count = 1; i < strbuf->size; i++) {
+        if (strbuf->data[i] == CR) {
+            count++;
+        }
+    }
+
+    return count;
+}
+
+void Strbuf_CopyLineNum(Strbuf *dst, const Strbuf *src, u32 linenum)
+{
+    Strbuf_Check(src);
+    Strbuf_Check(dst);
+
+    int i = 0;
+
+    if (linenum) {
+        for (i = 0; i < src->size; i++) {
+            if (src->data[i] == CR && --linenum == 0) {
+                i++;
+                break;
+            }
+        }
+    }
+
+    Strbuf_Clear(dst);
+
+    for ( ; i < src->size; i++) {
+        if (src->data[i] == CR) {
+            break;
+        }
+
+        Strbuf_Append(dst, src->data[i]);
+    }
+}
+
+void Strbuf_CopyChars(Strbuf *dst, const Charcode *src)
+{
+    Strbuf_Check(dst);
+
+    dst->size = 0;
+
+    while (*src != EOS) {
+        if (dst->size >= (dst->maxsize - 1)) {
+            GF_ASSERT(0);
+            break;
+        }
+
+        dst->data[dst->size++] = *src++;
+    }
+
+    dst->data[dst->size] = EOS;
+}
+
+void Strbuf_CopyNumChars(Strbuf *dst, const Charcode *src, u32 num)
+{
+    Strbuf_Check(dst);
+
+    if (num <= dst->maxsize) {
+        memcpy(dst->data, src, num * sizeof(Charcode));
+
+        u32 i;
+        for (i = 0; i < num; i++) {
+            if (dst->data[i] == EOS) {
+                break;
+            }
+        }
+
+        dst->size = i;
+        if (i == num) {
+            dst->data[num - 1] = EOS;
+        }
+
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+void Strbuf_ToChars(const Strbuf *src, Charcode *dst, u32 dstsize)
+{
+    Strbuf_Check(src);
+
+    if ((src->size + 1) <= dstsize) {
+        memcpy(dst, src->data, (src->size + 1) * sizeof(Charcode));
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+// Strbuf_GetData
+const Charcode* Strbuf_GetData(const Strbuf *src)
+{
+    Strbuf_Check(src);
+
+    return src->data;
+}
+
+void Strbuf_Concat(Strbuf *dst, const Strbuf *src)
+{
+    Strbuf_Check(dst);
+    Strbuf_Check(src);
+
+    if ((dst->size + src->size + 1) <= dst->maxsize) {
+        memcpy(dst->data + dst->size, src->data, (src->size + 1) * sizeof(Charcode));
+        dst->size += src->size;
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+// Strbuf_Append
+void Strbuf_Append(Strbuf *dst, Charcode c)
+{
+    Strbuf_Check(dst);
+
+    if ((dst->size + 1) < dst->maxsize) {
+        dst->data[dst->size++] = c;
+        dst->data[dst->size] = EOS;
+        return;
+    }
+
+    GF_ASSERT(0);
+}
```

Comment:
Not sure if we want to follow this style, but if we want to avoid mid-return statements, try the following two if they match. (discuss in Discord)
```c
    if ((dst->size + 1) < dst->maxsize) {
        dst->data[dst->size++] = c;
        dst->data[dst->size] = EOS;
    } else {
        GF_ASSERT(0);
    }
 ```
 
 ```c

    if ((dst->size + 1) >= dst->maxsize) {
        GF_ASSERT(0);
    } else {
        dst->data[dst->size++] = c;
        dst->data[dst->size] = EOS;        
    }
 ```

Reply:
We elected to allow mid-return statements.

Like before, the fewer nestings, the better.